pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -0,0 +1,223 @@
+            ack_ids = ack_ids[-2500:]
+            try:
+                await subscriber_client.acknowledge(subscription,
+                                                    ack_ids=ack_ids)
+            except Exception as e:
+                log.warning(
+                    'Ack request failed, better luck next batch', exc_info=e)
+                metrics_client.increment('pubsub.acker.batch.failed')
+                continue
+
+            metrics_client.histogram('pubsub.acker.batch', len(ack_ids))
+
+            ack_ids = []
+
+    async def _execute_callback(message: SubscriberMessage,
+                                callback: ApplicationHandler,
+                                ack_queue: 'asyncio.Queue[str]',
+                                metrics_client: MetricsAgent
+                                ) -> None:
+        try:
+            start = time.perf_counter()
+            await callback(message)
+            await ack_queue.put(message.ack_id)
+            metrics_client.increment('pubsub.consumer.succeeded')
+            metrics_client.histogram('pubsub.consumer.latency.runtime',
+                                     time.perf_counter() - start)
+        except Exception:
+            log.exception('Application callback raised an exception')
+            metrics_client.increment('pubsub.consumer.failed')
+
+    async def consumer(
+            message_queue: MessageQueue,
+            callback: ApplicationHandler,
+            ack_queue: 'asyncio.Queue[str]',
+            ack_deadline_cache: AckDeadlineCache,
+            consumer_pool_size: int,
+            metrics_client: MetricsAgent) -> None:
+        semaphore = asyncio.Semaphore(consumer_pool_size)
+        while True:
+            await semaphore.acquire()
+
+            message, pulled_at = await message_queue.get()
+
+            ack_deadline = await ack_deadline_cache.get()
+            if (time.perf_counter() - pulled_at) >= ack_deadline:
+                metrics_client.increment('pubsub.consumer.failfast')
+                message_queue.task_done()
+                semaphore.release()
+                continue
+
+            metrics_client.histogram(
+                'pubsub.consumer.latency.receive',
+                # publish_time is in UTC Zulu
+                # https://cloud.google.com/pubsub/docs/reference/rest/v1/PubsubMessage
+                time.time() - message.publish_time.timestamp())
+
+            task = asyncio.ensure_future(_execute_callback(
+                message,
+                callback,
+                ack_queue,
+                metrics_client,
+            ))
+            task.add_done_callback(lambda _f: semaphore.release())
+            message_queue.task_done()
+
+    async def producer(
+            subscription: str,
+            message_queue: MessageQueue,
+            subscriber_client: 'SubscriberClient',
+            max_messages: int,
+            metrics_client: MetricsAgent
+    ) -> None:
```

Comment:
nit: `) -> None:` in line with `async def` is not consistent with the rest of the declarations